src/cli/NativeMsgIface.cpp

```diff
@@ -0,0 +1,430 @@
+#if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0))
+    std::shared_ptr<QMetaObject::Connection> connection = std::make_shared<QMetaObject::Connection>();
+    return *connection = QObject::connect(sender, signal, context, [=](Args ...args) {
+        QObject::disconnect(*connection);
+        slot(std::forward<Args>(args)...);
+    });
+#else
+    return QObject::connect(sender, signal, context, slot, Qt::SingleShotConnection);
+#endif
+}
+
+// Little helper to print QSet<QString> and QList<QString> without the need to
+// convert them into a QStringList.
+template <typename T>
+QString join(QString glue, T const &list) {
+    QString out;
+    for (auto &&item : list) {
+        if (!out.isEmpty())
+            out += glue;
+        out += item;
+    }
+    return out;
+};
+
+}
+
+NativeMsgIface::NativeMsgIface(QObject * parent) :
+      QObject(parent)
+      , network_(this)
+      , settings_(this)
+      , models_(this, &settings_)
+      , operations_(0)
+      , die_(false)
+    {
+    qRegisterMetaType<std::shared_ptr<std::vector<char>>>("std::shared_ptr<std::vector<char>>");
+    
+    // Disable synchronisation with C style streams. That should make IO faster
+    std::ios_base::sync_with_stdio(false);
+
+    // Init the marian translation service:
+    marian::bergamot::AsyncService::Config serviceConfig;
+    serviceConfig.numWorkers = settings_.marianSettings().cpu_threads;
+    serviceConfig.cacheSize = settings_.marianSettings().translation_cache ? kTranslationCacheSize : 0;
+    service_ = std::make_shared<marian::bergamot::AsyncService>(serviceConfig);
+
+    // Pick up on network errors
+    // TODO: not sure if every request always returns on 1 error. fetchRemoteModels for example
+    // will cut short if a request is already in progress. If that one fails, the second request
+    // that got cut short will have no response, and stay pending.
+    connect(&network_, &Network::error, this, [&](QString err, QVariant data) {
+        if (data.canConvert<Request>())
+            writeError(data.value<Request>(), std::move(err));
+    });
+
+    connect(&network_, &Network::downloadComplete, this, [this](QFile *file, QString filename, QVariant data) {
+        models_.writeModel(file, filename);
+
+        if (data.canConvert<DownloadRequest>()) {
+            DownloadRequest request = data.value<DownloadRequest>();
+            writeResponse(request, QJsonObject{{"modelID", request.modelID}});
+        }
+    });
+
+    connect(this, &NativeMsgIface::emitJson, this, &NativeMsgIface::processJson);
+}
+
+void NativeMsgIface::run() {
+    iothread_ = std::thread([this](){
+        for (;;) {
+            // First part of the message: Find how long the input is. If that
+            // read fails, we're probably at EOF.
+            char len[4];
+            if (!std::cin.read(len, 4))
+                break;
+
+            int ilen = *reinterpret_cast<unsigned int *>(len);
```

Comment:
currently `_die` is never read, so we should probably remove it. I think That we can use `ilen == -1` as a signal that we want the native message client to die cleanly, as opposed to closing the `cin`.

Reply:
Closing cin is the common way to signal a downstream process to finish up? It is how pipes work. The native message server will stop only after it finished handling all pending requests after its stdin was closed. I have a test in `native_client.py` to test for this.